diff --git a/.htaccess b/.htaccess
index 4031da475c..5b87af6df3 100644
--- a/.htaccess
+++ b/.htaccess
@@ -60,6 +60,11 @@ AddEncoding gzip svgz
 <IfModule mod_rewrite.c>
   RewriteEngine on

+  # Redirect paths from the old site.
+  RewriteRule ^newsrooms/today/(.*)$ /$1 [R=301,L]
+  RewriteRule ^newsrooms/1/(.*)$ /$1 [R=301,L]
+  RewriteRule ^free-tags/(.*)$ /tag/$1 [R=301,L]
+
   # Set "protossl" to "s" if we were accessed via https://.  This is used later
   # if you enable "www." stripping or enforcement, in order to ensure that
   # you don't bounce between http and https.
